src/engine/runtime.js

```diff
@@ -172,29 +228,100 @@ Runtime.prototype.toggleScript = function (topBlockId) {
 };
 
 /**
- * Green flag, which stops currently running threads
- * and adds all top-level scripts that start with the green flag
+ * Run a function `f` for all scripts in a workspace.
+ * `f` will be called with two parameters:
+ *   -the top block ID of each script
+ *   -the opcode of that block, for convenience.
+ *   -fields on that block, for convenience.
+ * @param {!Function} f Function to call for each script.
+ * @param {Target=} opt_target Optionally, a target to restrict to.
  */
-Runtime.prototype.greenFlag = function () {
-    // Remove all existing threads
-    for (var i = 0; i < this.threads.length; i++) {
-        this._removeThread(this.threads[i]);
+Runtime.prototype.allScriptsDo = function (f, opt_target) {
+    var targets = this.targets;
+    if (opt_target) {
+        targets = [opt_target];
     }
-    // Add all top scripts with green flag
-    for (var t = 0; t < this.targets.length; t++) {
-        var target = this.targets[t];
+    for (var t = 0; t < targets.length; t++) {
+        var target = targets[t];
         var scripts = target.blocks.getScripts();
         for (var j = 0; j < scripts.length; j++) {
             var topBlock = scripts[j];
-            if (target.blocks.getBlock(topBlock).opcode ===
-                'event_whenflagclicked') {
-                this._pushThread(scripts[j]);
-            }
+            var topOpcode = target.blocks.getBlock(topBlock).opcode;
+            var topFields = target.blocks.getFields(topBlock);
+            f(topBlock, topOpcode, topFields);
         }
     }
 };
 
 /**
+ * Trigger all relevant hats.
+ * @param {!string} requestedHat Name of hat to trigger.
+ * @param {Object=} opt_matchFields Optionally, fields to match on the hat.
+ * @param {Target=} opt_target Optionally, a target to restrict to.
+ * @return {Array.<Thread>} List of threads started by this trigger.
+ */
+Runtime.prototype.triggerHats = function (requestedHat,
+    opt_matchFields, opt_target) {
+    var instance = this;
+    var newThreads = [];
+    // Consider all scripts, looking for hats named `requestedHat`.
+    this.allScriptsDo(function(topBlockId, topOpcode, topFields) {
+        if (topOpcode !== requestedHat) {
+            // Not the right hat.
+            return;
+        }
+        // Match any requested fields.
+        // For example: ensures that broadcasts match.
+        // This needs to happen before the block is evaluated
+        // (i.e., before the predicate can be run) because "broadcast and wait"
+        // needs to have a precise collection of triggered threads.
+        if (opt_matchFields) {
+            for (var matchField in opt_matchFields) {
+                if (topFields[matchField].value !==
+                    opt_matchFields[matchField]) {
+                    // Field mismatch.
+                    return;
+                }
+            }
+        }
+        if (instance._hats.hasOwnProperty(topOpcode)) {
```

Comment:
At first glance I thought `topOpcode` would be the first block under the hat rather than the hat itself, and it wasn't until this line that I realized I must be wrong, though I was quite confused about what the `if (topOpcode !== requestedHat)` above could mean!
Anyway, this might just be my brain slipping gears, but I do wonder if renaming the variable would make it more clear. Maybe something like `hatOpcode`? Somewhat related: `requestedHat` doesn't sound like an opcode to me (and your `@param` calls it a "name")... maybe `requestedHatOpcode` or something?